src/main/java/io/cdap/plugin/jms/sink/JMSRecordWriter.java

```diff
@@ -0,0 +1,114 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.jms.sink;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import io.cdap.plugin.jms.common.JMSConfig;
+import io.cdap.plugin.jms.common.JMSConnection;
+import io.cdap.plugin.jms.common.JMSMessageType;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import java.util.HashMap;
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.MapMessage;
+import javax.jms.Message;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+import javax.naming.Context;
+
+/**
+ * Record writer to produce messages to a JMS Topic/Queue.
+ */
+public class JMSRecordWriter extends RecordWriter<NullWritable, Text> {
+
+    private static final Gson GSON = new GsonBuilder().create();
+
+    private final JMSConfig jmsConfig;
+    private Connection connection;
+    private Session session;
+    private MessageProducer messageProducer;
+    private JMSConnection jmsConnection;
+
+    public JMSRecordWriter(TaskAttemptContext context) {
+        Configuration config = context.getConfiguration();
+        String configJson = config.get(JMSOutputFormatProvider.PROPERTY_CONFIG_JSON);
+        jmsConfig = GSON.fromJson(configJson, JMSConfig.class);
+        this.jmsConnection = new JMSConnection(jmsConfig);
+        establishConnection();
+    }
+
+    @Override
+    public void write(NullWritable key, Text text) {
+        try {
+            if (jmsConfig.getMessageType().equals(JMSMessageType.MAP.getName())) {
+                HashMap<String, Object> map = GSON.fromJson(text.toString(), new HashMap<String, Object>().getClass());
+                MapMessage mapMessage = session.createMapMessage();
+                map.forEach((k, v) -> {
+                    try {
+                        mapMessage.setObject(k, v);
+                    } catch (JMSException e) {
+                        e.printStackTrace();
+                    }
+                });
+                produceMessage(mapMessage);
+            } else {
+                TextMessage textMessage = session.createTextMessage();
+                textMessage.setText(text.toString());
+                produceMessage(textMessage);
+            }
+        } catch (JMSException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void close(TaskAttemptContext taskAttemptContext) {
+        try {
+            connection.stop();
+            session.close();
+            connection.close();
+        } catch (JMSException e) {
+            e.printStackTrace();
```

Comment:
log an error and remove the e.printStackTrace()

Reply:
Done